src/pkg/api/user.go

```diff
@@ -99,3 +99,180 @@ func (h UserHandler) Register(c *gin.Context) {
+// @Failure  401         "User unauthorized"
+// @Failure  500         "Is returned when an unexpected error occurs while processing a request"
+// @Router   /user/logout/ [post]
+func (h UserHandler) Logout(c *gin.Context) {
+	tokenString := c.Request.Header.Get("Authorization")
+	userID := c.Request.Header.Get("AuthUserId")
+
+	tokenString = h.service.ExtractToken(tokenString)
+
+	if tokenString == "" || userID == "" {
+		c.JSON(http.StatusUnauthorized, "unauthorized")
+		return
+	}
+
+	token, err := h.service.VerifyToken(userID, tokenString, false)
+	if err != nil {
+		c.JSON(http.StatusUnauthorized, "unauthorized")
+		return
+	}
+
+	metadata, err := h.service.ExtractTokenMetadata(token)
+	if err != nil {
+		c.JSON(http.StatusUnauthorized, "unauthorized")
+		return
+	}
+
+	delErr := h.service.DeleteToken(metadata.UUID)
+	if delErr != nil {
+		c.JSON(http.StatusUnauthorized, delErr.Error())
+		return
+	}
+
+	c.JSON(http.StatusOK, "Successfully logged out")
+}
+
+// RefreshToken
+// @Summary  Refresh JWT
+// @Description
+// @Tags     User
+// @Accept   json
+// @Produce  json
+// @Param    Authorization  header  string  true  "Bearer <JWT>"
+// @Param    AuthUserId  header  string  true  "UserId - UUID"
+// @Param    EhrSystemId  header  string  true  "The identifier of the system, typically a reverse domain identifier"
+// @Param    Request      body    model.JWT  true  "JWT"
+// @Success  201         {object}  model.JWT
+// @Failure  401         "User unauthorized"
+// @Failure  404         "User with ID not exist"
+// @Failure  422         "The request could not be understood by the server due to incorrect syntax. The client SHOULD NOT repeat the request without modifications."
+// @Failure  500         "Is returned when an unexpected error occurs while processing a request"
+// @Router   /user/refresh/ [post]
+func (h UserHandler) RefreshToken(c *gin.Context) {
+	userID := c.Request.Header.Get("AuthUserId")
+	tokenString := c.Request.Header.Get("Authorization")
+
+	tokenString = h.service.ExtractToken(tokenString)
+
+	if userID == "" || tokenString == "" {
+		c.JSON(http.StatusUnauthorized, "unauthorized")
+		return
+	}
+
+	var jwt model.JWT
+	if err := c.ShouldBindJSON(&jwt); err != nil {
+		c.JSON(http.StatusUnprocessableEntity, "Invalid json provided")
+		return
+	}
+
+	token, err := h.service.VerifyToken(userID, jwt.RefreshToken, true)
+	if err != nil {
+		c.JSON(http.StatusUnauthorized, "Refresh token expired")
+		return
+	}
+
+	metadata, err := h.service.ExtractTokenMetadata(token)
+	if err != nil {
+		c.JSON(http.StatusUnauthorized, "unauthorized")
+		return
+	}
+
+	delErr := h.service.DeleteToken(metadata.UUID)
+	if delErr != nil {
+		c.JSON(http.StatusUnauthorized, delErr.Error())
+		return
+	}
+
+	ts, err := h.service.CreateToken(metadata.UserID)
+	if err != nil {
+		c.JSON(http.StatusUnprocessableEntity, err.Error())
+		return
+	}
+
+	saveErr := h.service.CreateAuth(metadata.UserID, ts)
```

Comment:
Will a new refreshToken be created with each refresh?

Reply:
Yes. I saw this recommendation in several sources i.e. https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/#Refresh-Token-Rotation

Reply:
Then you have to delete the previously used token